### PR TITLE
Add Swagger support to API

### PR DESCRIPTION
**Pull request by lukepatrick**
This PR ports the API framework to a set of libraries that will generate Swagger/OpenAPI spec for consumers of the API. 

This addresses Issues #353 and #354, adding support and making the methods more discoverable.

### CODE REVIEW

File: brigade-api/cmd/brigade-api/main.go [hunk 1]
```diff
@@ -36,34 +193,43 @@ func main() {
...
+	config := restfulspec.Config{
+		WebServices: restful.RegisteredWebServices(), // you control what services are visible
+		APIPath:     "/apidocs.json",
+		PostBuildSwaggerObjectHandler: enrichSwaggerObject}
+	restful.DefaultContainer.Add(restfulspec.NewOpenAPIService(config))
 
-	j := server.Job()
-	rest.GET("/job/:id", j.Get)
-	rest.GET("/job/:id/logs", j.Logs)
+	// Optionally, you can install the Swagger Service which provides a nice Web UI on your REST API
```
Comment (adamreese) [inline brigade-api/cmd/brigade-api/main.go]: This comment can be removed

File: brigade-api/cmd/brigade-api/filter.go [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+				username = name
+			}
+		}
+		chain.ProcessFilter(req, resp)
+		ip, _, err := net.SplitHostPort(strings.TrimSpace(req.Request.RemoteAddr))
+		if err != nil {
+			return
+		}
+		var statusColor, methodColor, resetColor string
+		methodColor = colorForMethod(req.Request.Method)
```
Comment (adamreese) [inline brigade-api/cmd/brigade-api/filter.go]: Can you add a check for ANSI support.  Here is the one from gin https://github.com/gin-gonic/gin/blob/master/logger.go#L61-L63

File: brigade-api/cmd/brigade-api/filter.go [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+var (
+	green        = string([]byte{27, 91, 57, 55, 59, 52, 50, 109})
+	white        = string([]byte{27, 91, 57, 48, 59, 52, 55, 109})
+	yellow       = string([]byte{27, 91, 57, 55, 59, 52, 51, 109})
+	red          = string([]byte{27, 91, 57, 55, 59, 52, 49, 109})
+	blue         = string([]byte{27, 91, 57, 55, 59, 52, 52, 109})
+	magenta      = string([]byte{27, 91, 57, 55, 59, 52, 53, 109})
+	cyan         = string([]byte{27, 91, 57, 55, 59, 52, 54, 109})
+	reset        = string([]byte{27, 91, 48, 109})
+	disableColor = false
```
Comment (adamreese) [inline brigade-api/cmd/brigade-api/filter.go]: unused

File: brigade-api/cmd/brigade-api/filter.go [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+			req.Request.URL.RequestURI(),
+			req.Request.Proto,
+			statusColor, resp.StatusCode(), resetColor,
+			resp.ContentLength(),
+		)
+	}
+}
+
+// MeasureTime web-service (post-process) Filter (as a struct that defines a FilterFunction)
+func MeasureTime(req *restful.Request, resp *restful.Response, chain *restful.FilterChain) {
```
Comment (adamreese) [inline brigade-api/cmd/brigade-api/filter.go]: unused

File: brigade-api/cmd/brigade-api/main.go [hunk 5]
```diff
@@ -36,34 +193,43 @@ func main() {
...
+	// You need to download the Swagger HTML5 assets and change the FilePath location in the config below.
+	// Open http://localhost:8080/apidocs/?url=http://localhost:8080/apidocs.json
+	//http.Handle("/apidocs/", http.StripPrefix("/apidocs/", http.FileServer(http.Dir("/swagger-ui/dist"))))
 
-	router.GET("/healthz", api.Healthz)
+	// Optionally, you may need to enable CORS for the UI to work.
```
Comment (adamreese) [inline brigade-api/cmd/brigade-api/main.go]: We need to include CORS for Kashti

File: brigade-api/cmd/brigade-api/main.go [hunk 6]
```diff
@@ -27,6 +31,159 @@ func init() {
...
+		Returns(200, "OK", []byte{}).
+		Returns(404, "Not Found", nil))
+
+	return ws
+}
+
+func (ps projectService) WebService() *restful.WebService {
+	ws := new(restful.WebService)
+	p := ps.server.Project()
+	// rest.GET("/projects", p.List)
```
Comment (adamreese) [inline brigade-api/cmd/brigade-api/main.go]: These comments can be removed

**Comment by technosophos** [comment]
I'd really like to get this into the next Brigade release.

**Comment by lukepatrick** [comment]
updates pushed, let me know if anything else is needed

**Comment by technosophos** [comment]
Thanks! Marking this for review!

File: brigade-api/cmd/brigade-api/filter.go (hunk 3 above)
Comment (technosophos) [reply brigade-api/cmd/brigade-api/filter.go]: Looks like it is used now.

**Comment by adamreese** [comment]
Looks good, can you rebase and run `dep ensure`

**Comment by lukepatrick** [comment]
:ballot_box_with_check: rebased

**Pull request merged**